plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs

```diff
@@ -0,0 +1,142 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE TypeFamilies      #-}
+module Language.Plutus.Contract.Contract(
+      Contract
+    , emit
+    , waiting
+    , applyInput
+    , applyInputs
+    , drain
+    , outputs
+    , await
+    , result
+    , loopM
+    , foldMaybe
+    , firstOf
+    , select
+    , both
+    ) where
+
+import           Control.Applicative (liftA2)
+import           Control.Monad       ((>=>))
+import           Data.Bifunctor      (first)
+import           Data.List           (foldl')
+import           Data.Functor.Alt
+
+data Contract i o a =
+    Waiting (i -> Contract i o a)
+    | Emit o (Contract i o a) -- produce a 't' value
+    | Pure a
+    deriving (Functor)
+
+instance Alt (Contract i o) where
+    (<!>) = select
+
+-- The applicative instance parallelises the 'Waiting' operations
+instance Applicative (Contract i o) where
+    pure = Pure
+    cf <*> ca = case cf of
+        Emit t c -> Emit t (c <*> ca)
+        Pure f -> fmap f ca
+        Waiting f ->
+            case ca of
+                Emit t c -> Emit t (Waiting f <*> c)
+                Pure a' -> Waiting $ \i' -> fmap (\f' -> f' a') (f i')
+                Waiting f' ->
+                    Waiting $ \i' -> f i' <*> f' i'
+
+-- The monad instance sequentialises the 'Waiting' operations
+instance Monad (Contract i o) where
+    c >>= f = case c of
+        Waiting f' -> Waiting (f' >=> f)
+        Pure a     -> f a
+        Emit t c'  -> Emit t (c' >>= f)
+
+instance Semigroup a => Semigroup (Contract i o a) where
+    (<>) = liftA2 (<>)
+
+-- | Run both contracts and return the first one that finishes
+select :: Contract i o a -> Contract i o a -> Contract i o a
+select l r = 
```

Comment:
This function has a very similar shape to `<*>`, which makes me a bit suspicious, I wonder if it can be written in terms of it.